File: internal/format/format.go
// Package format turns numbers, dates and titles into the text haul prints and the names it gives files.
package format

import (
	"fmt"
	"math/rand/v2"
	"net/url"
	"path"
	"regexp"
	"strconv"
	"strings"
	"time"
	"unicode"
)

// FileSize is 512 bytes, 1.50 KB, 3.00 MB… with decimals after the point.
func FileSize(bytes float64, decimals int) string {
	b := max(0, bytes)
	const kb, mb, gb = 1024.0, 1024.0 * 1024, 1024.0 * 1024 * 1024
	switch {
	case b >= gb:
		return fmt.Sprintf("%.*f GB", decimals, b/gb)
	case b >= mb:
		return fmt.Sprintf("%.*f MB", decimals, b/mb)
	case b >= kb:
		return fmt.Sprintf("%.*f KB", decimals, b/kb)
	}
	return fmt.Sprintf("%d bytes", int64(b))
}

// Duration is 1h02m03s / 02m03s, or 01:02:03 when absolute.
func Duration(seconds int, absolute bool) string {
	h, m, s := seconds/3600, (seconds%3600)/60, seconds%60
	if absolute {
		return fmt.Sprintf("%02d:%02d:%02d", h, m, s)
	}
	if h == 0 {
		return fmt.Sprintf("%02dm%02ds", m, s)
	}
	return fmt.Sprintf("%dh%02dm%02ds", h, m, s)
}

// Timestamp renders unix seconds in local time with a Unicode date pattern (yyyy-MM-dd_HH-mm-ss). 0 renders as "null".
func Timestamp(ts int64, pattern string) string {
	if ts == 0 {
		return "null"
	}
	return FormatDate(time.Unix(ts, 0), pattern)
}

// FormatDate renders t with a Unicode date pattern: yyyy yy MM M dd d HH H hh h mm m ss s SSS a, and 'quoted' text.
func FormatDate(t time.Time, pattern string) string {
	var b strings.Builder
	rs := []rune(pattern)
	for i := 0; i < len(rs); {
		c := rs[i]
		if c == '\'' {
			j := i + 1
			for j < len(rs) && rs[j] != '\'' {
				j++
			}
			if j == i+1 {
				b.WriteRune('\'')
			} else {
				b.WriteString(string(rs[i+1 : j]))
			}
			i = j + 1
			continue
		}
		if !unicode.IsLetter(c) {
			b.WriteRune(c)
			i++
			continue
		}
		n := 1
		for i+n < len(rs) && rs[i+n] == c {
			n++
		}
		b.WriteString(dateField(t, c, n))
		i += n
	}
	return b.String()
}

func dateField(t time.Time, c rune, n int) string {
	pad := func(v int) string { return fmt.Sprintf("%0*d", n, v) }
	switch c {
	case 'y':
		if n == 2 {
			return fmt.Sprintf("%02d", t.Year()%100)
		}
		return pad(t.Year())
	case 'M':
		if n >= 3 {
			if n == 3 {
				return t.Month().String()[:3]
			}
			return t.Month().String()
		}
		return pad(int(t.Month()))
	case 'd':
		return pad(t.Day())
	case 'H':
		return pad(t.Hour())
	case 'h':
		h := t.Hour() % 12
		if h == 0 {
			h = 12
		}
		return pad(h)
	case 'm':
		return pad(t.Minute())
	case 's':
		return pad(t.Second())
	case 'S':
		ms := fmt.Sprintf("%09d", t.Nanosecond())
		if n > 9 {
			n = 9
		}
		return ms[:n]
	case 'a':
		if t.Hour() < 12 {
			return "AM"
		}
		return "PM"
	case 'E':
		if n >= 4 {
			return t.Weekday().String()
		}
		return t.Weekday().String()[:3]
	}
	return strings.Repeat(string(c), n)
}

// FFmpegCreationTime is ISO 8601 UTC with microseconds, the shape ffmpeg accepts for creation_time.
func FFmpegCreationTime(ts int64) string {
	return time.Unix(ts, 0).UTC().Format("2006-01-02T15:04:05.000000Z")
}

// ISO is ISO 8601 UTC with seconds.
func ISO(ts int64) string { return time.Unix(ts, 0).UTC().Format(time.RFC3339) }

// ParseDateTime reads `yyyy-MM-dd HH:mm:ss` (bilibili's pgc pub_time) as local time.
func ParseDateTime(text string) (int64, bool) {
	t, err := time.ParseInLocation("2006-01-02 15:04:05", text, time.Local)
	if err != nil {
		return 0, false
	}
	return t.Unix(), true
}

// ValidFileName replaces the characters no file system likes, slashes included (they separate directories).
func ValidFileName(input string) string {
	var b strings.Builder
	for _, r := range input {
		if r < 32 || strings.ContainsRune(`"<>|:*?\/`, r) {
			b.WriteByte('_')
		} else {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// CleanName is the cleanup every name component gets: invalid characters out, spaces and trailing dots trimmed.
func CleanName(input string) string {
	s := strings.TrimSpace(ValidFileName(input))
	s = strings.TrimRight(s, ".")
	return strings.TrimSpace(s)
}

// ChangeExtension swaps the extension of a path: a/b.mp4 → a/b.xml.
func ChangeExtension(p, ext string) string {
	return strings.TrimSuffix(p, path.Ext(p)) + "." + ext
}

var (
	decimalEntity = regexp.MustCompile(`&#(\d+);`)
	hexEntity     = regexp.MustCompile(`&#[xX]([0-9a-fA-F]+);`)
	namedEntities = strings.NewReplacer("&quot;", `"`, "&apos;", "'", "&lt;", "<", "&gt;", ">", "&nbsp;", " ", "&amp;", "&")
)

// UnescapeEntities turns &amp; &lt;… and numeric entities (&#26159; &#x4E2D;) back into characters.
func UnescapeEntities(text string) string {
	if !strings.Contains(text, "&") {
		return text
	}
	numeric := func(re *regexp.Regexp, base int) func(string) string {
		return func(m string) string {
			n, err := strconv.ParseInt(re.FindStringSubmatch(m)[1], base, 32)
			if err != nil || n < 0 || n > unicode.MaxRune {
				return m
			}
			return string(rune(n))
		}
	}
	text = decimalEntity.ReplaceAllStringFunc(text, numeric(decimalEntity, 10))
	text = hexEntity.ReplaceAllStringFunc(text, numeric(hexEntity, 16))
	return namedEntities.Replace(text)
}

// QueryValue is the raw value of name in a URL's query, or "".
func QueryValue(name, u string) string {
	q := strings.IndexByte(u, '?')
	if q < 0 {
		return ""
	}
	query := u[q+1:]
	if h := strings.IndexByte(query, '#'); h >= 0 {
		query = query[:h]
	}
	for _, pair := range strings.Split(query, "&") {
		k, v, ok := strings.Cut(pair, "=")
		if ok && k == name {
			return v
		}
	}
	return ""
}

// RandomString is length characters of the alphabet bilibili's device ids use.
func RandomString(length int) string {
	const chars = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_0123456789"
	b := make([]byte, length)
	for i := range b {
		b[i] = chars[rand.IntN(len(chars))]
	}
	return string(b)
}

// UnixSeconds is the current time in seconds, as text.
func UnixSeconds() string { return strconv.FormatInt(time.Now().Unix(), 10) }

// CompactTimestamp is yyyyMMddHHmmssSSS, used to seed the TV login fingerprint.
func CompactTimestamp() string { return FormatDate(time.Now(), "yyyyMMddHHmmssSSS") }

// PercentEncode escapes everything but letters, digits and -._~.
func PercentEncode(s string) string {
	return strings.ReplaceAll(url.QueryEscape(s), "+", "%20")
}

File: Sources/HaulCore/Core/Format.swift
import Foundation

public enum Format {
    public static func fileSize(_ bytes: Double, decimals: Int = 2) -> String {
        let b = max(0, bytes)
        let kb = 1024.0, mb = kb * 1024, gb = mb * 1024
        let f = "%.\(decimals)f"
        if b >= gb { return String(format: f + " GB", b / gb) }
        if b >= mb { return String(format: f + " MB", b / mb) }
        if b >= kb { return String(format: f + " KB", b / kb) }
        return "\(Int(b)) bytes"
    }

    /// `1h02m03s` / `02m03s`, or `01:02:03` when `absolute`.
    public static func duration(_ seconds: Int, absolute: Bool = false) -> String {
        let h = seconds / 3600
        let m = (seconds % 3600) / 60
        let s = seconds % 60
        if absolute { return String(format: "%02d:%02d:%02d", h, m, s) }
        return h == 0 ? String(format: "%02dm%02ds", m, s) : String(format: "%dh%02dm%02ds", h, m, s)
    }

    /// Unix seconds → local time in a Unicode date pattern (e.g. `yyyy-MM-dd_HH-mm-ss`). `0` renders as `null`, like the original.
    public static func timestamp(_ ts: Int64, pattern: String) -> String {
        guard ts != 0 else { return "null" }
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = .current
        f.dateFormat = pattern
        return f.string(from: Date(timeIntervalSince1970: TimeInterval(ts)))
    }

    /// ISO-8601 UTC with microseconds, the shape ffmpeg accepts for `creation_time`.
    public static func ffmpegCreationTime(_ ts: Int64) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'"
        return f.string(from: Date(timeIntervalSince1970: TimeInterval(ts)))
    }

    /// Parse `yyyy-MM-dd HH:mm:ss` (bilibili's pgc pub_time) as local time.
    public static func parseDateTime(_ text: String) -> Int64? {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = .current
        f.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return f.date(from: text).map { Int64($0.timeIntervalSince1970) }
    }

    private static let invalidFileChars: Set<Character> = {
        var s: Set<Character> = ["\"", "<", ">", "|", ":", "*", "?", "\\", "/"]
        for i in 0..<32 { s.insert(Character(UnicodeScalar(UInt8(i)))) }
        return s
    }()

    /// Replace characters no filesystem likes. Slashes are always replaced; they are directory separators, not name characters.
    public static func validFileName(_ input: String, replacement: String = "_") -> String {
        var out = ""
        out.reserveCapacity(input.count)
        for ch in input { out += invalidFileChars.contains(ch) ? replacement : String(ch) }
        return out
    }

    /// Title cleanup applied to every name component: strip, drop trailing dots, strip again.
    public static func cleanName(_ input: String) -> String {
        var s = validFileName(input).trimmingCharacters(in: .whitespaces)
        while s.hasSuffix(".") { s.removeLast() }
        return s.trimmingCharacters(in: .whitespaces)
    }

    /// `&amp;`, `&lt;`… and numeric entities (`&#26159;`, `&#x4E2D;`) back to characters.
    public static func unescapeEntities(_ text: String) -> String {
        guard text.contains("&") else { return text }
        var s = text
        s = s.replacing(/&#(\d+);/) { m in Int(m.1).flatMap { Unicode.Scalar($0) }.map { String(Character($0)) } ?? String(m.0) }
        s = s.replacing(/&#[xX]([0-9a-fA-F]+);/) { m in Int(m.1, radix: 16).flatMap { Unicode.Scalar($0) }.map { String(Character($0)) } ?? String(m.0) }
        for (entity, char) in [("&quot;", "\""), ("&apos;", "'"), ("&lt;", "<"), ("&gt;", ">"), ("&nbsp;", " "), ("&amp;", "&")] {
            s = s.replacingOccurrences(of: entity, with: char)
        }
        return s
    }

    public static func queryValue(_ name: String, in url: String) -> String {
        guard let q = url.firstIndex(of: "?") else { return "" }
        let query = url[url.index(after: q)...].split(separator: "#").first ?? ""
        for pair in query.split(separator: "&") {
            let kv = pair.split(separator: "=", maxSplits: 1)
            if kv.count == 2, kv[0] == name { return String(kv[1]) }
        }
        return ""
    }

    public static func randomString(_ length: Int) -> String {
        let chars = Array("ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_0123456789")
        return String((0..<length).map { _ in chars.randomElement()! })
    }

    public static func unixSeconds() -> String { String(Int64(Date().timeIntervalSince1970)) }
    public static func unixMillis() -> String { String(Int64(Date().timeIntervalSince1970 * 1000)) }

    /// `yyyyMMddHHmmssfff` — used to seed the TV login fingerprint.
    public static func compactTimestamp() -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyyMMddHHmmssSSS"
        return f.string(from: Date())
    }

    public static func percentEncode(_ s: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return s.addingPercentEncoding(withAllowedCharacters: allowed) ?? s
    }
}
